logstash-core/src/test/java/org/logstash/plugins/pipeline/PipelineBusTest.java

```diff
@@ -307,6 +316,56 @@ public void whenInBlockingModeInputsShutdownLast() throws InterruptedException {
         assertThat(bus.getAddressState(address)).isNotPresent();
     }
 
+    @Test
+    public void blockingShutdownDeadlock() throws InterruptedException {
+        final ExecutorService executor = Executors.newFixedThreadPool(10);
+        try {
+            for (int i = 0; i < 100; i++) {
+                bus.registerSender(output, addresses);
+                bus.listen(input, address);
+                bus.setBlockOnUnlisten(true);
+
+                // we use a CountDownLatch to increase the likelihood
+                // of simultaneous execution
```

Comment:
With the countdown latch in place, this test reliably finds the preexisting deadlock in <5 iterations, and without it we can reliably find it in <40 iterations. The test runs 100 iterations just to be safe.

Reply:
Brilliant! I was curious if there was anything else for getting threads to wait on a particular point and found `CyclicBarrier` https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/CyclicBarrier.html 

I tried running several times with this diff:
```diff
diff --git a/logstash-core/src/test/java/org/logstash/plugins/pipeline/PipelineBusTest.java b/logstash-core/src/test/java/org/logstash/plugins/pipeline/PipelineBusTest.java
index 268ed8d09..1dbaddaab 100644
--- a/logstash-core/src/test/java/org/logstash/plugins/pipeline/PipelineBusTest.java
+++ b/logstash-core/src/test/java/org/logstash/plugins/pipeline/PipelineBusTest.java
@@ -39,6 +39,7 @@ import java.util.List;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
@@ -325,17 +326,14 @@ public class PipelineBusTest {
                 bus.listen(input, address);
                 bus.setBlockOnUnlisten(true);

-                // we use a CountDownLatch to increase the likelihood
-                // of simultaneous execution
-                final CountDownLatch startLatch = new CountDownLatch(2);
+                // Synchronize when unlisten/unregisterSender start
+                final CyclicBarrier startBarrier = new CyclicBarrier(2);
                 final CompletableFuture<Void> unlistenFuture = CompletableFuture.runAsync(asRunnable(() -> {
-                    startLatch.countDown();
-                    startLatch.await();
+                    startBarrier.await();
                     bus.unlisten(input, address);
                 }), executor);
                 final CompletableFuture<Void> unregisterFuture = CompletableFuture.runAsync(asRunnable(() -> {
-                    startLatch.countDown();
-                    startLatch.await();
+                    startBarrier.await();
                     bus.unregisterSender(output, addresses);
                 }), executor);
```
Regardless of whether I was using `CountdownLatch` or `CyclicBarrier` I could not really see a meaningful difference in (very good with either option) performance.